map/framework.cpp

```diff
@@ -3313,3 +3313,70 @@ void Framework::OnPowerSchemeChanged(power_management::Scheme const actualScheme
   if (actualScheme == power_management::Scheme::EconomyMaximum && GetTrafficManager().IsEnabled())
     GetTrafficManager().SetEnabled(false);
 }
+
+bool Framework::ShouldShowProducts() const
+{
+  auto const connectionStatus = Platform::ConnectionStatus();
+  if (connectionStatus == Platform::EConnectionType::CONNECTION_NONE)
+    return false;
+
+  std::string donateUrl;
+  if (!settings::Get(settings::kDonateUrl, donateUrl)) // donation is disabled
+    return false;
+
+  if (!storage::IsPointCoveredByDownloadedMaps(GetCurrentPlacePageInfo().GetMercator(), m_storage, * m_infoGetter))
+    return false;
+
+  #ifdef DEBUG
+  uint32_t constexpr kFirstLaunchTimeout = 60 * 3; // 3 min because it may takes some time to download the map after the first launch in debug
+  uint32_t constexpr kLastProductCloseTimeout = 10;
+  uint32_t constexpr kLastProductSelectTimeout = 20;
+  #else
+  uint32_t constexpr kFirstLaunchTimeout = 60 * 60 * 3; // 3 hours
+  uint32_t constexpr kLastProductCloseTimeout = 60 * 60 * 24  * 14; // 14 days
+  uint32_t constexpr kLastProductSelectTimeout = 180 * 24 * 60 * 60; // 180 days
+  #endif
+
+  auto const now = base::SecondsSinceEpoch();
+
+  if (m_usageStats.GetFirstLaunchTime() + kFirstLaunchTimeout > now)
+    return false;
+
+  uint64_t popupCloseTime;
+  if (!settings::Get(settings::kPlacePageProductsCloseTime, popupCloseTime))
+    settings::Set(settings::kPlacePageProductsCloseTime, 0);
+  uint64_t productSelectTime;
+  if (!settings::Get(settings::kPlacePageProductsSelectedTime, productSelectTime))
+    settings::Set(settings::kPlacePageProductsSelectedTime, 0);
```

Comment:
- I suspect that should be:
uint64_t popupCloseTime = 0;
uint64_t productSelectTime = 0;
- What is the reason to make settings::Set(..., 0) ?

Reply:
I suggest the explicit set the 0 if there is no such fileds in settings
```cpp
  uint64_t popupCloseTime;
  if (!settings::Get(settings::kPlacePageProductsCloseTime, popupCloseTime))
    popupCloseTime = 0; // popup was never shown
  uint64_t productSelectTime;
  if (!settings::Get(settings::kPlacePageProductsSelectedTime, productSelectTime))
    productSelectTime = 0; // product was never selected

  bool const closePopupTimeoutExpired = popupCloseTime == 0 || popupCloseTime + kLastProductCloseTimeout < now;
  bool const productSelectTimeoutExpired = productSelectTime == 0 || productSelectTime + kLastProductSelectTimeout < now;
  if (closePopupTimeoutExpired && productSelectTimeoutExpired)
    return true;
```